File: src/obsidian_wiki_langgraph/core/link_engine.py
"""Bidirectional link engine for Obsidian wiki-links.

Handles [[wiki-link]] injection, extraction, backlink queries, and graph
construction.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from pathlib import Path

from .utils import extract_title, walk_markdown_files


# ---------------------------------------------------------------------------
# Data structures
# ---------------------------------------------------------------------------

@dataclass
class GraphNode:
    path: str
    title: str | None


@dataclass
class GraphEdge:
    source: str
    target: str


@dataclass
class Graph:
    nodes: list[GraphNode] = field(default_factory=list)
    edges: list[GraphEdge] = field(default_factory=list)


# ---------------------------------------------------------------------------
# LinkEngine
# ---------------------------------------------------------------------------

class LinkEngine:
    """Bidirectional link engine for an Obsidian vault.

    Responsible for [[wiki-link]] injection, parsing, reverse lookups and
    graph construction.
    """

    def __init__(self, vault_path: str | Path) -> None:
        self.vault_path = Path(vault_path).resolve()
        self._title_cache: dict[str, str] = {}  # normalized_title -> path
        self._cache_valid: bool = False

    # --- Link injection ---------------------------------------------------

    def refresh_title_cache(self) -> None:
        """Scan the vault and rebuild the title cache."""
        self._title_cache.clear()
        files = walk_markdown_files(self.vault_path, self.vault_path)

        for file in files:
            abs_path = self.vault_path / file
            content = abs_path.read_text(encoding="utf-8")
            title = extract_title(content)

            # basename without .md
            basename = file.removesuffix(".md")
            name_only = Path(file).stem

            if title:
                self._title_cache[title.lower()] = file
            self._title_cache[name_only.lower()] = file
            # Also register full relative path
            self._title_cache[basename.lower()] = file

        self._cache_valid = True

    def get_note_titles(self, folder: str | None = None) -> list[dict]:
        """Return a list of {path, title} dicts for all (or folder-scoped) notes."""
        if not self._cache_valid:
            self.refresh_title_cache()

        if folder:
            files = walk_markdown_files(self.vault_path / folder, self.vault_path)
        else:
            files = walk_markdown_files(self.vault_path, self.vault_path)

        result: list[dict] = []
        for f in files:
            abs_path = self.vault_path / f
            content = abs_path.read_text(encoding="utf-8")
            result.append({"path": f, "title": extract_title(content)})
        return result

    def inject_links(self, content: str) -> str:
        """Auto-inject [[wiki-links]] into note content.

        Rules:
         1. Skip text already inside [[...]]
         2. Skip frontmatter region
         3. Match known note titles -> [[path]]
         4. Match dates YYYY-MM-DD -> [[310-Daily/YYYY-MM-DD]]
        """
        if not self._cache_valid:
            self.refresh_title_cache()

        # Separate frontmatter
        fm_part = ""
        body_part = content

        if content.startswith("---"):
            end_index = content.find("---", 3)
            if end_index != -1:
                fm_part = content[: end_index + 3]
                body_part = content[end_index + 3 :]

        # Collect all known titles, sorted by key length descending (longest first)
        sorted_titles = sorted(
            [
                {
                    "key": key,
                    "file_path": file_path,
                    "display": file_path.removesuffix(".md"),
                }
                for key, file_path in self._title_cache.items()
            ],
            key=lambda x: len(x["key"]),
            reverse=True,
        )

        result = body_part

        # Inject date links: YYYY-MM-DD not already inside [[...]]
        result = re.sub(
            r"(?<!\[\[)(\d{4}-\d{2}-\d{2})(?!\]\])",
            lambda m: f"[[310-Daily/{m.group(1)}]]",
            result,
        )

        # Inject known title links
        for entry in sorted_titles:
            key = entry["key"]
            display = entry["display"]
            if len(key) < 2:
                continue  # Skip very short titles

            # Escape regex special characters
            escaped = re.escape(key)
            # Match text not preceded by [[ or / and not followed by ]]
            pattern = rf"(?<!\[/)\b({escaped})\b(?!\])"
            result = re.sub(
                pattern,
                lambda m, _display=display: f"[[{_display}]]",
                result,
                flags=re.IGNORECASE,
            )

        # Clean nested links: [[[[a]]]] -> [[a]]
        result = re.sub(r"\[\[\[([^\]]+)\]\]\]", r"[[\1]]", result)

        return fm_part + result

    # --- Link extraction --------------------------------------------------

    @staticmethod
    def extract_links(content: str) -> list[str]:
        """Extract all [[wiki-link]] targets from content.

        Handles both [[target]] and [[target|alias]] forms. Strips leading
        ../ segments and removes .md suffixes. Returns unique links.
        """
        links: list[str] = []
        regex = r"\[\[([^\]|]+?)(?:\|[^\]]+?)?\]\]"
        for match in re.finditer(regex, content):
            target = match.group(1).strip()
            # Strip leading ../
            while target.startswith("../"):
                target = target[3:]
            if target.endswith(".md"):
                target = target.removesuffix(".md")
            links.append(target)

        return list(dict.fromkeys(links))  # unique, preserving order

    def get_outlinks(self, note_path: str) -> list[str]:
        """Get all outgoing links from a note."""
        abs_path = self.vault_path / note_path
        if not abs_path.exists():
            return []

        content = abs_path.read_text(encoding="utf-8")
        return LinkEngine.extract_links(content)

    def get_backlinks(self, note_path: str) -> list[dict]:
        """Get all backlinks pointing to *note_path*.

        Returns a list of {source, context} dicts.
        """
        files = walk_markdown_files(self.vault_path, self.vault_path)
        results: list[dict] = []

        # Possible reference forms for the target note
        basename = Path(note_path).stem
        targets = [
            note_path.removesuffix(".md"),
            basename,
            f"/{basename}",
            f"/{note_path.removesuffix('.md')}",
        ]

        for file in files:
            if file == note_path:
                continue
            abs_path = self.vault_path / file
            content = abs_path.read_text(encoding="utf-8")
            lines = content.split("\n")

            for line in lines:
                links = LinkEngine.extract_links(line)
                has_match = any(
                    link.lower() == t.lower() or link.lower().endswith(t.lower())
                    for link in links
                    for t in targets
                )

                if has_match:
                    results.append({
                        "source": file,
                        "context": line.strip()[:200],
                    })
                    break  # Only first matching line per file

        return results

    # --- Graph construction -----------------------------------------------

    def build_graph(self) -> Graph:
        """Build the full bidirectional link graph."""
        files = walk_markdown_files(self.vault_path, self.vault_path)
        nodes: list[GraphNode] = []
        edges: list[GraphEdge] = []

        for file in files:
            abs_path = self.vault_path / file
            content = abs_path.read_text(encoding="utf-8")

            nodes.append(GraphNode(path=file, title=extract_title(content)))

            outlinks = LinkEngine.extract_links(content)
            for target in outlinks:
                edges.append(GraphEdge(source=file, target=target))

        return Graph(nodes=nodes, edges=edges)

File: src/link-engine.ts
import * as fs from "fs";
import * as path from "path";
import { walkMarkdownFiles, extractTitle } from "./utils.js";

export interface GraphNode {
  path: string;
  title: string | null;
}

export interface GraphEdge {
  source: string;
  target: string;
}

export interface Graph {
  nodes: GraphNode[];
  edges: GraphEdge[];
}

/**
 * 双向链接引擎
 * 负责 [[wiki-link]] 的注入、解析、反向查询和图谱构建
 */
export class LinkEngine {
  private vaultPath: string;
  private titleCache: Map<string, string> = new Map(); // normalizedTitle → path
  private cacheValid = false;

  constructor(vaultPath: string) {
    this.vaultPath = path.resolve(vaultPath);
  }

  // ─── 链接注入 ──────────────────────────────────

  /**
   * 刷新标题缓存：扫描 vault 所有笔记标题
   */
  refreshTitleCache(): void {
    this.titleCache.clear();
    const files = walkMarkdownFiles(this.vaultPath, this.vaultPath);

    for (const file of files) {
      const absPath = path.join(this.vaultPath, file);
      const content = fs.readFileSync(absPath, "utf-8");
      const title = extractTitle(content);

      // 用文件名（去掉 .md）作为 key
      const basename = file.replace(/\.md$/, "");
      const nameOnly = path.basename(file, ".md");

      if (title) {
        this.titleCache.set(title.toLowerCase(), file);
      }
      this.titleCache.set(nameOnly.toLowerCase(), file);
      // 也注册完整路径
      this.titleCache.set(basename.toLowerCase(), file);
    }

    this.cacheValid = true;
  }

  /**
   * 获取所有笔记标题列表
   */
  getNoteTitles(folder?: string): Array<{ path: string; title: string | null }> {
    if (!this.cacheValid) this.refreshTitleCache();

    const files = folder
      ? walkMarkdownFiles(path.join(this.vaultPath, folder), this.vaultPath)
      : walkMarkdownFiles(this.vaultPath, this.vaultPath);

    return files.map((f) => {
      const absPath = path.join(this.vaultPath, f);
      const content = fs.readFileSync(absPath, "utf-8");
      return { path: f, title: extractTitle(content) };
    });
  }

  /**
   * 对内容自动注入 [[双向链接]]
   * 规则：
   *  1. 跳过已经在 [[...]] 内的文本
   *  2. 跳过 frontmatter 区域
   *  3. 匹配已知笔记标题 → [[路径]]
   *  4. 匹配日期 YYYY-MM-DD → [[310-Daily/YYYY-MM-DD]]
   */
  injectLinks(content: string): string {
    if (!this.cacheValid) this.refreshTitleCache();

    // 分离 frontmatter
    let fmPart = "";
    let bodyPart = content;

    if (content.startsWith("---")) {
      const endIndex = content.indexOf("---", 3);
      if (endIndex !== -1) {
        fmPart = content.slice(0, endIndex + 3);
        bodyPart = content.slice(endIndex + 3);
      }
    }

    // 收集所有已知标题，按长度降序排列（长标题优先匹配）
    const sortedTitles = [...this.titleCache.entries()]
      .map(([key, filePath]) => ({
        key,
        filePath,
        display: filePath.replace(/\.md$/, ""),
      }))
      .sort((a, b) => b.key.length - a.key.length);

    let result = bodyPart;

    // 注入日期链接
    result = result.replace(
      /(?<!\[\[)(\d{4}-\d{2}-\d{2})(?!\]\])/g,
      (_, date) => {
        // 检查是否已经在链接中
        return `[[310-Daily/${date}]]`;
      }
    );

    // 注入已知标题链接
    for (const { key, display } of sortedTitles) {
      if (key.length < 2) continue; // 跳过太短的标题
      // 转义正则特殊字符
      const escaped = key.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
      // 匹配不在 [[...]] 内的文本
      const regex = new RegExp(`(?<![\\[/])\\b(${escaped})\\b(?![\\]])`, "gi");
      result = result.replace(regex, () => `[[${display}]]`);
    }

    // 清理嵌套链接：[[[[a]]]] → [[a]]
    result = result.replace(/\[\[\[([^\]]+)\]\]\]/g, "[[$1]]");

    return fmPart + result;
  }

  // ─── 链接解析 ──────────────────────────────────

  /**
   * 从内容中提取所有 [[wiki-link]] 目标
   */
  static extractLinks(content: string): string[] {
    const links: string[] = [];
    const regex = /\[\[([^\]|]+?)(?:\|[^\]]+?)?\]\]/g;
    let match: RegExpExecArray | null;

    while ((match = regex.exec(content)) !== null) {
      let target = match[1].trim();
      // 去掉开头的 ../
      while (target.startsWith("../")) {
        target = target.slice(3);
      }
      if (!target.endsWith(".md")) {
        // 尝试解析为路径
        links.push(target);
      } else {
        links.push(target.replace(/\.md$/, ""));
      }
    }

    return [...new Set(links)];
  }

  /**
   * 获取某笔记的所有正向链接（outgoing links）
   */
  getOutlinks(notePath: string): string[] {
    const absPath = path.join(this.vaultPath, notePath);
    if (!fs.existsSync(absPath)) return [];

    const content = fs.readFileSync(absPath, "utf-8");
    return LinkEngine.extractLinks(content);
  }

  /**
   * 获取某笔记的所有反向链接（backlinks）
   */
  getBacklinks(notePath: string): Array<{
    source: string;
    context: string;
  }> {
    const files = walkMarkdownFiles(this.vaultPath, this.vaultPath);
    const results: Array<{ source: string; context: string }> = [];

    // 目标笔记的各种可能引用形式
    const basename = path.basename(notePath, ".md");
    const targets = [
      notePath.replace(/\.md$/, ""),
      basename,
      `/${basename}`,
      `/${notePath.replace(/\.md$/, "")}`,
    ];

    for (const file of files) {
      if (file === notePath) continue;
      const absPath = path.join(this.vaultPath, file);
      const content = fs.readFileSync(absPath, "utf-8");
      const lines = content.split("\n");

      for (const line of lines) {
        const links = LinkEngine.extractLinks(line);
        const hasMatch = links.some((link) =>
          targets.some(
            (t) => link.toLowerCase() === t.toLowerCase()
              || link.toLowerCase().endsWith(t.toLowerCase()),
          )
        );

        if (hasMatch) {
          results.push({
            source: file,
            context: line.trim().slice(0, 200),
          });
          break; // 每个文件只取第一个匹配行
        }
      }
    }

    return results;
  }

  // ─── 图谱构建 ──────────────────────────────────

  /**
   * 构建完整的双向链接图谱
   */
  buildGraph(): Graph {
    const files = walkMarkdownFiles(this.vaultPath, this.vaultPath);
    const nodes: GraphNode[] = [];
    const edges: GraphEdge[] = [];

    for (const file of files) {
      const absPath = path.join(this.vaultPath, file);
      const content = fs.readFileSync(absPath, "utf-8");

      nodes.push({
        path: file,
        title: extractTitle(content),
      });

      const outlinks = LinkEngine.extractLinks(content);
      for (const target of outlinks) {
        edges.push({ source: file, target });
      }
    }

    return { nodes, edges };
  }
}
